src/main/java/org/blockchain_innovation/factom/identiy/did/request/CreateFactomDidRequest.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    public String[] getTags() {
+        return this.tags;
+    }
+
+    public DIDVersion getDidVersion() {
+        return this.didVersion;
+    }
+
+    public FactomDidContent toFactomDidContent() {
+        String chainId = new CreateFactomDIDEntry(this.didVersion, null, this.nonce, this.tags).getChainId();
+        String did = this.networkName == null ? "did:factom:" + chainId : "did:factom:" + this.networkName + ':' + chainId;
+        return new FactomDidContent()
+                .didMethodVersion(DidMethodVersion.fromValue(didVersion.getSchemaVersion()))
+                .didKey(this.didKey.stream().map(key -> key.toDidKey(did))
+                        .collect(Collectors.toList()))
+                .managementKey(this.managementKey.stream().map(key -> key.toManagementKey(did))
+                        .collect(Collectors.toList()))
+                .service(this.service.stream().map(didService -> didService.toService(did))
+                        .collect(Collectors.toList()));
+    }
+
+    public static final class Builder {
+        private DIDVersion didVersion;
+        private List<CreateKeyRequest> managementKey;
+        private List<CreateKeyRequest> didKey;
+        private List<CreateServiceRequest> service;
+        private List<String> tags;
+        private String nonce;
+        private String networkName;
+
+        public Builder() {
+        }
+
+        public Builder didVersion(DIDVersion didVersion) {
+            this.didVersion = didVersion;
+            return this;
+        }
+
+        public Builder managementKey(List<CreateKeyRequest> managementKey) {
+            this.managementKey = managementKey;
+            return this;
+        }
+
+        public Builder didKey(List<CreateKeyRequest> didKey) {
+            this.didKey = didKey;
+            return this;
+        }
+
+        public Builder service(List<CreateServiceRequest> service) {
+            this.service = service;
+            return this;
+        }
+
+        public Builder tag(String tag) {
+            if (this.tags == null) {
+                this.tags = new ArrayList<>(Arrays.asList(tag));
+            } else {
+                this.tags.add(tag);
+            }
+            return this;
+        }
+
+        public Builder nonce(String nonce) {
+            this.nonce = nonce;
+            return this;
+        }
+
+        public Builder networkName(String networkName) {
+            this.networkName = networkName;
+            return this;
+        }
+
+        public CreateFactomDidRequest build() throws IncompleteRequestException {
+            if (this.service == null) {
+                this.service = Collections.emptyList();
+            }
+            this.assertComplete();
+            return new CreateFactomDidRequest(didVersion, networkName, managementKey, didKey, service, nonce, tags.toArray(new String[0]));
+        }
+
+        private void assertComplete() throws IncompleteRequestException {
```

Comment:
Suggest to put the validation in an interface across all requests together with other possible candidates and if there are multiple candidate methods user a template method design pattern where applicable

Reply:
I'm not sure I understand what you mean here:

>  an interface across all requests together with other possible candidates

I can make an interface that is extended by each request, but the other requests do not require validation as the errors are handled by the existing rules (also I'm not sure what else should go in this interface other than the validation method). This class only has extra validation because `toFactomDidContent` will expect a non-null value. 

Either we could remove the validation here and just create an empty list for any missing values (like we do with the optional services), which would forward the validation responsibility onto the existing rules, or we could add a validator for each request.

Reply:
Ah okay, I expected the validation the be quite common across the request objects I guess.